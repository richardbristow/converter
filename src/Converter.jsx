import React, { Component } from 'react';
import Dropdown from './Dropdown';
import conversionTypes from './static/conversionTypes';
import conversionUnits from './static/conversionUnits';

class Converter extends Component {
  constructor(props) {
    super(props);

    this.state = { conversionType: 'Length' };
    this.handleChange = this.handleChange.bind(this);
  }

  handleChange(event) {
    this.setState({ conversionType: event.target.value });
  }

  render() {
    return (
      <div>
        <Dropdown
          conversionTypes={conversionTypes}
          value={this.state.conversionType}
          handleChange={this.handleChange}
        />
        <br />
        <br />
        <input type="text" />
        <Dropdown
          conversionTypes={conversionUnits[`${this.state.conversionType}`]}
          value={this.state.conversionType}
          handleChange={this.handleChange}
        />
        <br />
        <br />
        <input type="text" />
        <Dropdown
          conversionTypes={conversionUnits[`${this.state.conversionType}`]}
          value={this.state.conversionType}
          handleChange={this.handleChange}
        />
      </div>
    );
  }
}

export default Converter;
