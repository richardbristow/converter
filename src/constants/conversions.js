import length from './units/length';
import surfaceArea from './units/surfaceArea';

const conversions = {
  length,
  surfaceArea,
  // volume,
  // liquidVolume,
  // angles,
  // time,
  // frequency,
  // mass,
  // electricCurrent,
  // temperature,
  // amountOfSubstance,
  // luminousIntensity,
  // currency,
  // force,
  // energy,
  // power,
  // pressure,
  // electricityAndMagnetism,
  // binary,
};

export default conversions;
