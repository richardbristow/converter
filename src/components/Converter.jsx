import React, { Component } from 'react';
import Dropdown from './Dropdown';
import Input from './Input';
import conversions from '../constants/conversions';
import currentUnits from '../utils/currentUnits';
import { leftToRight, rightToLeft, tryConvert } from '../utils/calculator';

class Converter extends Component {
  constructor(props) {
    super(props);

    this.state = {
      conversionType: conversions.length.mathName,
      lengthUnitLeft: conversions.length.initialUnitLeft,
      lengthUnitRight: conversions.length.initialUnitRight,
      inputLeft: '',
      inputRight: '',
    };

    this.handleChange = this.handleChange.bind(this);
  }

  handleChange({ target }) {
    const { name, value } = target;
    const unitNameLeft = `${value}UnitLeft`;
    const unitNameRight = `${value}UnitRight`;
    const unitValueLeft = this.state[unitNameLeft];
    const unitValueRight = this.state[unitNameRight];
    if (name === 'conversionType' && !unitValueLeft && !unitValueRight) {
      this.setState({
        [name]: value,
        [unitNameLeft]: conversions[value].initialUnitLeft,
        [unitNameRight]: conversions[value].initialUnitRight,
      });
    } else {
      this.setState({ [name]: value });
    }
  }

  render() {
    const { inputLeft, inputRight, conversionType } = this.state;
    const {
      unitNameLeft, unitNameRight, unitValueLeft, unitValueRight,
    } = currentUnits(conversionType, this.state);
    return (
      <div>
        <Dropdown
          className="conversionType"
          name="conversionType"
          options={Object.keys(conversions).map((type) => {
            const reformattedObject = {
              displayName: conversions[type].displayName,
              mathName: conversions[type].mathName,
            };
            return reformattedObject;
          })}
          value={conversionType}
          handleChange={this.handleChange}
        />
        <br />
        <br />
        <Input name="inputLeft" value={inputLeft} handleChange={this.handleChange} />
        <Dropdown
          className="unitLeft"
          name={unitNameLeft}
          options={conversions[conversionType].units}
          value={unitValueLeft}
          handleChange={this.handleChange}
        />
        <br />
        <br />
        <Input name="inputRight" value={inputRight} handleChange={this.handleChange} />
        <Dropdown
          className="unitRight"
          name={unitNameRight}
          options={conversions[conversionType].units}
          value={unitValueRight}
          handleChange={this.handleChange}
        />
      </div>
    );
  }
}

export default Converter;
