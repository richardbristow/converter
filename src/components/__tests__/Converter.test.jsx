import React from 'react';
import { shallow } from 'enzyme';
import Converter from '../Converter';
import conversions from '../../constants/conversions';

describe('Converter', () => {
  it('should render 3 dropdown menus', () => {
    const wrapper = shallow(<Converter />);
    expect(wrapper.find('Dropdown')).toHaveLength(3);
  });

  it('should mount dropdowns with default values', () => {
    const {
      mathName: defaultType,
      initialUnitLeft: defaultUnitLeft,
      initialUnitRight: defaultUnitRight,
    } = conversions.length;

    const wrapper = shallow(<Converter />);
    expect(wrapper.find('Dropdown.conversionType').props().value).toBe(defaultType);
    expect(wrapper.find('Dropdown.unitLeft').props().value).toBe(defaultUnitLeft);
    expect(wrapper.find('Dropdown.unitRight').props().value).toBe(defaultUnitRight);
  });
});


describe('handleChange', () => {
  it('should initialise state to unit default values when the conversionType is changed', () => {
    const wrapper = shallow(<Converter />);
    const inst = wrapper.instance();
    inst.handleChange({
      target: { value: 'surfaceArea', name: 'conversionType' },
    });
    expect(inst.state.surfaceAreaUnitLeft).toBe('m2');
    expect(inst.state.surfaceAreaUnitRight).toBe('sqin');
    expect(inst.state).toMatchSnapshot();
  });

  it('should add {name: "value"} as a key value pair to the state', () => {
    const wrapper = shallow(<Converter />);
    const inst = wrapper.instance();
    inst.handleChange({
      target: { value: 'VALUE', name: 'NAME' },
    });
    expect(Object.keys(inst.state)).toContain('NAME');
    expect(Object.values(inst.state)).toContain('VALUE');
    expect(inst.state.NAME).toBe('VALUE');
  });

  it('should update existing unit values with new values', () => {
    const wrapper = shallow(<Converter />);
    const inst = wrapper.instance();
    inst.handleChange({
      target: { value: 'OldValueLeft', name: 'OldUnitLeft' },
    });
    if (inst.state.OldUnitLeft) {
      inst.handleChange({
        target: { value: 'NewValueLeft', name: 'OldUnitLeft' },
      });
    }
    expect(inst.state.OldUnitLeft).toBe('NewValueLeft');
  });
});
