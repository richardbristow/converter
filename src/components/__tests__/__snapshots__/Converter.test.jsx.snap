// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`handleChange should initialise state to unit default values when the conversionType is changed 1`] = `
Object {
  "conversionType": "surfaceArea",
  "inputLeft": "",
  "inputRight": "",
  "lengthUnitLeft": "meter",
  "lengthUnitRight": "inch",
  "surfaceAreaUnitLeft": "m2",
  "surfaceAreaUnitRight": "sqin",
}
`;
